source/extensions/filters/http/api_key_auth/api_key_auth.cc

```diff
@@ -0,0 +1,172 @@
+#include "source/extensions/filters/http/api_key_auth/api_key_auth.h"
+
+#include <openssl/sha.h>
+
+#include "envoy/http/header_map.h"
+
+#include "source/common/common/base64.h"
+#include "source/common/http/header_utility.h"
+#include "source/common/http/headers.h"
+#include "source/common/http/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace ApiKeyAuth {
+
+ApiKeyAuthConfig::ApiKeyAuthConfig(const ApiKeyAuthProto& proto_config)
+    : key_sources_(proto_config.key_sources()) {
+  if (!proto_config.has_credentials()) {
```

Comment:
If there are no credentials but the route has provide one, then the filter still could works. This provide the biggest flexibility to let users to combine the filter config and route config.

And even there is no any credentials finally at both filter and route, the filter will handle that case correctly by return a 401 response because we cannot authenticate anyone at that case.